### PR TITLE
Added support for hierarchical and scoped attributes.

### PR DESCRIPTION
**Pull request by irinaepshteyn**
Signed-off-by: Irina irina.epshteyn@ge.com

### CODE REVIEW

File: acs-integration-tests/pom.xml [hunk 1]
```diff
@@ -675,6 +674,109 @@
         </profile>
 
         <profile>
+            <id>public-titan</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
```
Comment (sanjeevchopra) [inline acs-integration-tests/pom.xml]: this profile should not be active by default, yet. ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/PrivilegeManagementServiceImpl.java [hunk 2]
```diff
@@ -290,6 +297,23 @@ public BaseSubject getBySubjectIdentifier(final String subjectIdentifier) {
         return subject;
     }
 
+    @Transactional(readOnly = true)
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/PrivilegeManagementServiceImpl.java:301]: Line 289: (above) getBySubjectIdentifier  - this method needs to use a different implementation on the graph repository, such that it _only_ returns the attributes explicitly set on the subject. Currently this will return all inherited attributes as well. ✅

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 3]
```diff
@@ -141,8 +157,81 @@ public void testGetByZoneAndResourceIdentifier() {
...
+    public void testGetByZoneAndResourceIdentifier3LevelHierarchical() {
+        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(0L));
+        String resourceIdentifier = persist3LevelHierarchicalResource1toZone1().getResourceIdentifier();
+        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(3L));
+
+        ResourceEntity resource = this.resourceRepository.getByZoneAndResourceIdentifier(TEST_ZONE_1,
+                resourceIdentifier);
+        assertThat(resource.getAttributes().contains(ATTR_SITE_0), equalTo(true));
+        assertThat(resource.getAttributes().contains(ATTR_TYPE_1), equalTo(true));
+        assertThat(resource.getAttributes().contains(ATTR_CLASSIFICATION_0), equalTo(true));
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: this test will be even more easier to follow, if the attributes being asserted were passed in to the utility method at line 180  , where the hierarchy is created.

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java [hunk 4]
```diff
@@ -242,15 +333,90 @@ public static String getVertexStringPropertyOrNull(final Vertex vertex, final St
...
+            Assert.isTrue(!traversal.hasNext(),
+                    String.format("There are two entities with the same %s.", getEntityIdKey()));
+            return entity;
+        } finally {
+            getGraph().tx().commit();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public void searchAttributes(final E entity, final Vertex vertex) {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java]: is there a unit test which directly tests this method and the one below ?  at this level we can test cases which are not in the Resource/Subject repository tests
- vertex with outbound edges to 2 other vertices  (subject or resource with 2 parents)
- others ?

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java [hunk 5]
```diff
@@ -242,15 +333,90 @@ public static String getVertexStringPropertyOrNull(final Vertex vertex, final St
...
+
+        getGraph().traversal().V(vertex.id()).has(ATTRIBUTES_PROPERTY_KEY).emit()
+                .repeat(outE().has(SCOPE_PROPERTY_KEY, test(elementOf(), scopes)).otherV().simplePath()
+                        .has(ATTRIBUTES_PROPERTY_KEY))
+                .until(eq(null)).limit(this.traversalLimit).values(ATTRIBUTES_PROPERTY_KEY).toStream().forEach(it -> {
+                    attributes.addAll(JSON_UTILS.deserialize((String) it, Set.class, Attribute.class));
+                });
+
+        entity.setAttributes(attributes);
+        entity.setAttributesAsJson(JSON_UTILS.serialize(attributes));
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java]: Why do we need to keep the json serialized as a string ?

File: model/src/main/java/com/ge/predix/acs/model/Attribute.java [hunk 6]
```diff
@@ -41,7 +41,7 @@
     private List<Attribute> scopes;
 
     public Attribute() {
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/model/Attribute.java:40]: scopes need to be removed from this class ?    Also ScopedAttribute class is not used, I think - the class and its references need to be removed.  ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepository.java [hunk 7]
```diff
@@ -54,7 +48,7 @@ String getRelationshipKey() {
     }
 
     @Override
-    void upsertEntityVertex(final SubjectEntity entity, final Vertex vertex) {
+    void updateVertexProperties(final SubjectEntity entity, final Vertex vertex) {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepository.java:55]: (created story for this: US126731)
line: 62 It will be more accurate to store each individual attribute as a property of the subject vertex.  Will also allow us to traverse the graph based on attributes. ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java [hunk 8]
```diff
@@ -242,15 +333,90 @@ public static String getVertexStringPropertyOrNull(final Vertex vertex, final St
...
+        } finally {
+            getGraph().tx().commit();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public void searchAttributes(final E entity, final Vertex vertex) {
+        Set<Attribute> attributes = new HashSet<>();
+        getGraph().traversal().V(vertex.id()).has(ATTRIBUTES_PROPERTY_KEY).emit()
+                .repeat(out().simplePath().has(ATTRIBUTES_PROPERTY_KEY)).until(eq(null)).limit(this.traversalLimit)
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java]: Need a way to indicate to the user, when the attributes were limited.  Maybe this should fail instead of limiting the attributes - because it will affect the access decision.

File: model/src/main/java/com/ge/predix/acs/rest/Parent.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+ */
+@ApiModel(description = "Represents a source of inherited attributes.")
+public class Parent {
+    @ApiModelProperty(
+            value = "The unique identifier of the parent, e.g. \"tom@ge.com\" or \"/site/sanramon\".",
+            required = true)
+    private String identifier;
+
+    @ApiModelProperty(
+            value = "Restrictions on when a subject inherits parent attributes based on the resource accessed,"
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/rest/Parent.java:26]: Clarify behavior when 'scopes' set has more than one attribute.  Also need unit tests for this usecase.

For now, we could throw NotSupportedException when more than one scope attribute is set.  And implement this as a separate story.

File: model/src/main/java/com/ge/predix/acs/rest/BaseResource.java [hunk 10]
```diff
@@ -38,6 +38,8 @@
     private String resourceIdentifier;
     @ApiModelProperty(value = "The list of attribute values being assigned", required = true)
     private Set<Attribute> attributes = Collections.emptySet();
+    @ApiModelProperty(value = "A set that identifies resources whose attributes also apply to this resource")
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/rest/BaseResource.java]: ..to this reource.., during policy evaluation.  ✅

File: model/src/main/java/com/ge/predix/acs/rest/BaseSubject.java [hunk 11]
```diff
@@ -40,6 +40,9 @@
     @ApiModelProperty(value = "The list of attribute values being assigned", required = true)
     private Set<Attribute> attributes = Collections.emptySet();
 
+    @ApiModelProperty(value = "A set that identifies subjects whose attributes also apply to this resource")
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/rest/BaseSubject.java]: ..to this subject.., during policy evaluation. ✅

File: model/src/main/java/com/ge/predix/acs/rest/PolicyEvaluationRequestV1.java [hunk 12]
```diff
@@ -55,10 +57,19 @@ public void setSubjectIdentifier(final String subjectIdentifier) {
         this.subjectIdentifier = subjectIdentifier;
     }
 
+    @ApiModelProperty(value = "Supplemental resource attributes provided by the requestor", required = true)
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/rest/PolicyEvaluationRequestV1.java]: required=false ✅

File: model/src/main/java/com/ge/predix/acs/rest/PolicyEvaluationRequestV1.java [hunk 13]
```diff
@@ -55,10 +57,19 @@ public void setSubjectIdentifier(final String subjectIdentifier) {
...
+
+    public void setResourceAttributes(final List<Attribute> resourceAttributes) {
+        this.resourceAttributes = resourceAttributes;
+    }
+
     /**
      * @return the subjectAttributes
      */
-    @ApiModelProperty(value = "The subject attributes", required = true)
+    @ApiModelProperty(value = "Supplemental subject attributes provided by the requestor", required = true)
```
Comment (sanjeevchopra) [inline model/src/main/java/com/ge/predix/acs/rest/PolicyEvaluationRequestV1.java]: required=false  ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/PrivilegeManagementServiceImpl.java (hunk 2 above)
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/PrivilegeManagementServiceImpl.java]: getSubjects() method never refers to subjectScopedAccessRepository. It has to use this when titan is enabled.

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java [hunk 14]
```diff
@@ -188,20 +211,88 @@ public E findOne(final Long id) {
...
+        if (!traversal.hasNext()) {
+            throw new IllegalStateException(String.format("No parent exists in zone '%s' with '%s' value of '%s'.",
+                    entity.getZone().getName(), getEntityIdKey(), parent.getIdentifier()));
+        }
+        if (parent.getScopes().isEmpty()) {
+            vertex.addEdge(PARENT_EDGE_LABEL, traversal.next());
+            return;
+        }
+        for (Attribute scope : parent.getScopes()) {
+            vertex.addEdge(PARENT_EDGE_LABEL, traversal.next()).property(SCOPE_PROPERTY_KEY,
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java]: addEdge before loop ? and run edge.property in a loop for each scope. ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/ParentEntity.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
+package com.ge.predix.acs.privilege.management.dao;
+
+import java.util.Set;
+
+import com.ge.predix.acs.model.Attribute;
+
+public class ParentEntity {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/ParentEntity.java:7]: Confirm that this class is needed. Its only used in one method which is not used.

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolver.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        this.privilegeService = privilegeService;
+        this.subjectIdentifier = subjectIdentifier;
+        if (null == supplementalSubjectAttributes) {
+            this.supplementalSubjectAttributes = Collections.emptySet();
+        } else {
+            this.supplementalSubjectAttributes = supplementalSubjectAttributes;
+        }
+    }
+
+    public SubjectAttributeResolverResult getResult(final String resourceIdentifier, final Set<Attribute> scopes) {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolver.java]: return Set<Attribute> instead of SubjectAttributeResolverResult, and remove SubjectAttributeResolverResult  ✅

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolver.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+        Set<Attribute> subjectAttributes = this.subjectAttributeMap.get(resourceIdentifier);
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolver.java]: use this.subjectIdentifier as the key. remove param resourceIdentifier ✅

File: service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java [hunk 18]
```diff
@@ -242,15 +333,90 @@ public static String getVertexStringPropertyOrNull(final Vertex vertex, final St
...
+                .repeat(out().simplePath().has(ATTRIBUTES_PROPERTY_KEY)).until(eq(null)).limit(this.traversalLimit)
+                .values(ATTRIBUTES_PROPERTY_KEY).toStream().forEach(it -> {
+                    attributes.addAll(JSON_UTILS.deserialize((String) it, Set.class, Attribute.class));
+                });
+        entity.setAttributes(attributes);
+        entity.setAttributesAsJson(JSON_UTILS.serialize(attributes));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void searchAttributesWithScopes(final E entity, final Vertex vertex, final Set<Attribute> scopes) {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/privilege/management/dao/GraphGenericRepository.java]: add unit test : vertex with 2 outbound edges, where one edge has a scope property.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 19]
```diff
@@ -17,19 +31,20 @@
...
 
 import com.ge.predix.acs.config.GraphConfig;
+import com.ge.predix.acs.rest.Parent;
+import com.ge.predix.acs.utils.JsonUtils;
 import com.ge.predix.acs.zone.management.dao.ZoneEntity;
 import com.thinkaurelius.titan.core.SchemaViolationException;
 import com.thinkaurelius.titan.core.TitanFactory;
 
 public class GraphResourceRepositoryTest {
-
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:30]: - Delete tests should lookup the vertex after delete and assert that is not found.  (in addition to existing asserts on count)
- find tests should assert entity inserted equals entity returned by find, in addition to count asserts
- testGetByZoneAndResourceIdentifier should compare all properties in ResourceEntity. Maybe add a equals method in the entity.

File: service/src/test/java/com/ge/predix/acs/testutils/Constants.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
+package com.ge.predix.acs.testutils;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.ge.predix.acs.model.Attribute;
+
+public class Constants {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/testutils/Constants.java]: https://github.com/predix/acs/commit/8e578aa6d97ca901f651f5115016c00bff3bbba9  - renamed constants to improve understandibility.  ✅

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvalWithGraphDbControllerIT.java [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+import com.ge.predix.acs.testutils.MockMvcContext;
+import com.ge.predix.acs.testutils.MockSecurityContext;
+import com.ge.predix.acs.testutils.TestUtils;
+import com.ge.predix.acs.utils.JsonUtils;
+import com.ge.predix.acs.zone.management.ZoneService;
+
+@WebAppConfiguration
+@ContextConfiguration("classpath:controller-tests-context.xml")
+@ActiveProfiles(profiles = { "h2", "public", "simple-cache", "titan" })
+public class PolicyEvalWithGraphDbControllerIT extends AbstractTestNGSpringContextTests {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvalWithGraphDbControllerIT.java:69]: - Consider re-using the subject/resource graphs in GraphSubjectRepositoryTest & GraphResourceRepositoryTest, for ease of understanding. Same code can be used, and any nodes added in this test, if needed.

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeMgmtWithGraphDbControllerIT.java [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+import com.ge.predix.acs.testutils.MockMvcContext;
+import com.ge.predix.acs.testutils.MockSecurityContext;
+import com.ge.predix.acs.testutils.TestUtils;
+import com.ge.predix.acs.utils.JsonUtils;
+import com.ge.predix.acs.zone.management.ZoneService;
+
+@WebAppConfiguration
+@ContextConfiguration("classpath:controller-tests-context.xml")
+@ActiveProfiles(profiles = { "h2", "public", "simple-cache", "titan" })
+public class ResourcePrivilegeMgmtWithGraphDbControllerIT extends AbstractTestNGSpringContextTests {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeMgmtWithGraphDbControllerIT.java]: This does not need to be an integration test. If there are additional testcases from GraphResourceRepositoryTest, can we add them as unit tests ?

**Comment by sanjeevchopra** [comment]
[@Bitty9 , Sanjeev]  PR review complete.

**Comment by sanjeevchopra** [comment]
saw warnings in test logs about titan indices (below). Lets review, in case we are missing some indices.

> [Z: C:] 2016-06-08 17:45:32 INFO [main] o.s.t.c.t.TransactionContext [TransactionContext.java:136] Rolled back transaction for test context [DefaultTestContext@102fcc80 testClass = PrivilegeManagementServiceImplTest, testInstance = com.ge.predix.acs.privilege.management.PrivilegeManagementServiceImplTest@45815ffc, testMethod = testUpdateSubject@PrivilegeManagementServiceImplTest, testException = [null], mergedContextConfiguration = [MergedContextConfiguration@2182cd7c testClass = PrivilegeManagementServiceImplTest, locations = '{}', classes = '{class com.ge.predix.acs.request.context.AcsRequestContextHolder, class com.ge.predix.acs.policy.evaluation.cache.HystrixPolicyEvaluationCacheCircuitBreaker, class com.ge.predix.acs.config.InMemoryDataSourceConfig, class com.ge.predix.acs.policy.evaluation.cache.InMemoryPolicyEvaluationCache, class com.ge.predix.acs.privilege.management.PrivilegeManagementServiceImpl, class com.ge.predix.acs.SpringSecurityPolicyContextResolver, class com.ge.predix.acs.zone.resolver.SpringSecurityZoneResolver, class com.ge.predix.acs.zone.management.ZoneServiceImpl}', contextInitializerClasses = '[]', activeProfiles = '{h2, public, simple-cache}', propertySourceLocations = '{}', propertySourceProperties = '{}', contextLoader = 'org.springframework.test.context.support.DelegatingSmartContextLoader', parent = [null]]].
> [Z: C:] 2016-06-08 17:45:32 INFO [main] o.s.c.s.GenericApplicationContext [AbstractApplicationContext.java:578] Refreshing org.springframework.context.support.GenericApplicationContext@3bc1ed53: startup date [Wed Jun 08 17:45:32 PDT 2016]; root of context hierarchy
> [Z: C:] 2016-06-08 17:45:32 WARN [main] o.s.c.a.ConfigurationClassEnhancer [ConfigurationClassEnhancer.java:347] @Bean method Config.properties is non-static and returns an object assignable to Spring's BeanFactoryPostProcessor interface. This will result in a failure to process annotations such as @Autowired, @Resource and @PostConstruct within the method's declaring @Configuration class. Add the 'static' modifier to this method to avoid these container lifecycle issues; see @Bean javadoc for complete details.
> [Z: C:] 2016-06-08 17:45:32 INFO [main] o.s.c.s.DefaultLifecycleProcessor [DefaultLifecycleProcessor.java:341] Starting beans in phase 2147483647
> [Z: C:] 2016-06-08 17:45:33 WARN [hystrix-HystrixPolicyEvaluationCacheCircuitBreaker-3] c.g.p.a.p.e.c.HystrixPolicyEvaluationCacheCircuitBreaker [HystrixPolicyEvaluationCacheCircuitBreaker.java:48] Executing circuit breaker fallback method for cache 'set' method.
> [Z: C:] 2016-06-08 17:45:33 WARN [hystrix-HystrixPolicyEvaluationCacheCircuitBreaker-4] c.g.p.a.p.e.c.HystrixPolicyEvaluationCacheCircuitBreaker [HystrixPolicyEvaluationCacheCircuitBreaker.java:34] Executing circuit breaker fallback method for cache 'get' method.
> [Z: C:] 2016-06-08 17:45:33 WARN [hystrix-HystrixPolicyEvaluationCacheCircuitBreaker-7] c.g.p.a.p.e.c.HystrixPolicyEvaluationCacheCircuitBreaker [HystrixPolicyEvaluationCacheCircuitBreaker.java:61] Executing circuit breaker fallback method for cache 'setResourceTranslation' method.
> [Z: C:] 2016-06-08 17:45:33 WARN [hystrix-HystrixPolicyEvaluationCacheCircuitBreaker-8] c.g.p.a.p.e.c.HystrixPolicyEvaluationCacheCircuitBreaker [HystrixPolicyEvaluationCacheCircuitBreaker.java:76] Executing circuit breaker fallback method for cache 'setResourceTranslations' method.
> [Z: C:] 2016-06-08 17:45:34 INFO [main] c.t.t.g.c.GraphDatabaseConfiguration [GraphDatabaseConfiguration.java:1358] Set default timestamp provider MICRO
> [Z: C:] 2016-06-08 17:45:34 INFO [main] c.t.t.g.c.GraphDatabaseConfiguration [GraphDatabaseConfiguration.java:1518] Generated unique-instance-id=032759403901-sjc1jenkins071
> [Z: C:] 2016-06-08 17:45:34 INFO [main] c.t.t.d.Backend [Backend.java:176] Initiated backend operations thread pool of size 16
> [Z: C:] 2016-06-08 17:45:34 INFO [main] c.t.t.d.l.k.KCVSLog [KCVSLog.java:730] Loaded unidentified ReadMarker start time 2016-06-09T00:45:34.388Z into com.thinkaurelius.titan.diskstorage.log.kcvs.KCVSLog$MessagePuller@73ebf7cc
> [Z: C:] 2016-06-08 17:45:35 INFO [main] c.t.t.g.d.m.GraphIndexStatusWatcher [GraphIndexStatusWatcher.java:69] All 1 key(s) on index byZone have status ENABLED
> [Z: C:] 2016-06-08 17:45:36 INFO [main] c.t.t.g.d.m.GraphIndexStatusWatcher [GraphIndexStatusWatcher.java:69] All 2 key(s) on index byZoneAndResourceUnique have status ENABLED
> [Z: C:] 2016-06-08 17:45:36 INFO [main] c.t.t.g.d.m.GraphIndexStatusWatcher [GraphIndexStatusWatcher.java:69] All 2 key(s) on index byZoneAndSubjectUnique have status ENABLED
> [Z: C:] 2016-06-08 17:45:36 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:37 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:37 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:38 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:39 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:39 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:41 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:41 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:41 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:41 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [(resourceId = /x-files/drive)]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [(resourceId = /x-files/drive)]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [()]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 WARN [main] c.t.t.g.t.StandardTitanTx [StandardTitanTx.java:1262] Query requires iterating over all vertices [(resourceId = /x-files/drive)]. For better performance, use indexes
> [Z: C:] 2016-06-08 17:45:42 INFO [main] o.s.c.s.GenericApplicationContext [AbstractApplicationContext.java:578] Refreshing org.springframework.context.support.GenericApplicationContext@5440942: startup date [Wed Jun 08 17:45:42 PDT 2016]; root of context hierarchy

**Comment by sanjeevchopra** [comment]
All unit and integration tests need to run with "titan" profile to get full coverage against a graph repository. Currently, only the new integration tests are being run in this profile.

in progress:  removed titan profile from GraphConfig. Unit tests need to include "GraphBeanDefinitionRegistryPostProcessor.class, GraphConfig.class"  in ContextConfiguration to use graph db.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java [hunk 23]
```diff
@@ -18,12 +34,14 @@
 import org.testng.annotations.Test;
 
 import com.ge.predix.acs.config.GraphConfig;
+import com.ge.predix.acs.model.Attribute;
+import com.ge.predix.acs.rest.Parent;
+import com.ge.predix.acs.utils.JsonUtils;
 import com.ge.predix.acs.zone.management.dao.ZoneEntity;
 import com.thinkaurelius.titan.core.TitanFactory;
 
 public class GraphSubjectRepositoryTest {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java:45]: This class should include the hierarchical tests in resource repository test.  We should try to reuse the tests with some parameterization.
